packages/core/database/lib/entity-manager/relations-orderer.js

```diff
@@ -1,6 +1,106 @@
 'use strict';
 
 const _ = require('lodash/fp');
+const { InvalidRelationError } = require('../errors');
+/**
+ * When connecting relations, the order you connect them matters.
+ *
+ * Example, if you connect the following relations:
+ *   { id: 5, position: { before: 1 } }
+ *   { id: 1, position: { before: 2 } }
+ *   { id: 2, position: { end: true } }
+ *
+ * Going through the connect array, id 5 has to be connected before id 1,
+ * so the order of id5 = id1 - 1. But the order value of id 1 is unknown.
+ * The only way to know the order of id 1 is to connect it first.
+ *
+ * This function makes sure the relations are connected in the right order:
+ *   { id: 2, position: { end: true } }
+ *   { id: 1, position: { before: 2 } }
+ *   { id: 5, position: { before: 1 } }
+ *
+ */
+const sortConnectArray = (connectArr, initialArr = [], strictSort = true) => {
+  const sortedConnect = [];
+  // Boolean to know if we have to recalculate the order of the relations
+  let needsSorting = false;
+  // Map to validate if relation is already in sortedConnect or DB.
+  const relInArray = initialArr.reduce((acc, rel) => ({ ...acc, [rel.id]: true }), {});
+  // Map to store the first index where a relation id is connected
+  const firstSeen = {};
+  // Map to validate if connect relation has already been computed
+  const computedIdx = {};
+
+  connectArr.forEach((rel, idx) => {
+    // If adjacent relation is not in the database or seen yet in the connect array
+    // then we need to sort the connect array
+    const adjacentRelId = rel.position?.before || rel.position?.after;
+    if (!relInArray[adjacentRelId] && !firstSeen[adjacentRelId]) needsSorting = true;
+    // Populate firstSeen
+    if (!(rel.id in firstSeen)) firstSeen[rel.id] = idx;
+  });
+
+  // If we don't need to sort the connect array, we can return it as is
+  if (!needsSorting) return connectArr;
+
+  // Add relation to sortedConnect and mark it as seen
+  const pushRelation = (rel) => {
+    sortedConnect.push(rel);
+    relInArray[rel.id] = true;
+  };
```

Comment:
I think you can avoid short names to [avoid mental mapping](https://github.com/ryanmcdermott/clean-code-javascript#avoid-mental-mapping)

Reply:
sure, but is it hard to read rel in the context of relations?
I think there is a tradeof between mental mapping and readability. 

I would find it hard to understand if it was `x` or `r` instead of `rel`. In my eyes rel is explicit enough in this context. Do you think it's better to always write the full word?

Reply:
I agree that `rel` is better than an `x` variable. I do think that it still needs some mental computation to make the link between `rel` and `relation` (a bit the same as `i` and `index`, we are used to short index with `i`).
I think the optimum maximum length for a variable name is around 12 characters, IMO. Then use shorts (like `relInArray`).